### PR TITLE
Confirm concurrent HTTP/2 requests with empty flow-control window.

### PR DESCRIPTION
**Pull request by dave-r12**
I'm not fully convinced this is worth merging. I just wanted to attempt to reproduce the steps in https://github.com/square/okhttp/issues/3915.

### CODE REVIEW

File: okhttp-tests/src/test/java/okhttp3/internal/http2/HttpOverHttp2Test.java [hunk 1]
```diff
@@ -367,6 +363,38 @@ private static OkHttpClient buildHttp2Client() {
...
+        .url(server.url("/"))
+        .build());
+    Response response2 = call2.execute();
+    assertEquals(200, response2.code());
+
+    // Close the response body. This should discard the buffered data and update the connection
+    // flow-control window.
+    response1.close();
+
+    assertEquals("abc", response2.body().string());
```
Comment (yschimke) [inline okhttp-tests/src/test/java/okhttp3/internal/http2/HttpOverHttp2Test.java]: Can we make the test stronger e.g. assert nothing can be read until after response1.close? Perhaps using inputStream.available()?

Comment (dave-r12) [reply okhttp-tests/src/test/java/okhttp3/internal/http2/HttpOverHttp2Test.java]: I'm not sure if there is a good way of checking that. It looks like `available()` will always return 0 unless you trigger a read. From my perspective, this test is proving that the second request will proceed as usual given the actions from the first request.

**Comment by natez0r** [comment]
@dave-r12 this looks like it could be the issue I am seeing. My second request always fails when attempting to take the headers.

For completeness' sake, do you have any pointers on where i could throw a breakpoint in the OKHTTP code to confirm this in my test app?

**Comment by dave-r12** [comment]
@natez0r currently this test passes! So there is likely something different in your environment.

> For completeness' sake, do you have any pointers on where i could throw a breakpoint in the OKHTTP code to confirm this in my test app?

A possible first step: could you turn on frame logging and paste in the output?

**Comment by natez0r** [comment]
When I wrote that comment, I had checked out your patch and that test failed for me:

> [INFO] -------------------------------------------------------
[INFO] Running okhttp3.internal.http2.HttpOverHttp2Test
[ERROR] Tests run: 90, Failures: 0, Errors: 2, Skipped: 5, Time elapsed: 38.732 s \<\<\< FAILURE! - in okhttp3.internal.http2.HttpOverHttp2Test
[ERROR] concurrentRequestWithEmptyFlowControlWindow[h2_prior_knowledge](okhttp3.internal.http2.HttpOverHttp2Test)  Time elapsed: 10.092 s  \<\<\< ERROR!
java.net.SocketTimeoutException: timeout
	at okhttp3.internal.http2.HttpOverHttp2Test.concurrentRequestWithEmptyFlowControlWindow(HttpOverHttp2Test.java:388)
[ERROR] concurrentRequestWithEmptyFlowControlWindow[h2](okhttp3.internal.http2.HttpOverHttp2Test)  Time elapsed: 10.206 s  \<\<\< ERROR!
java.net.SocketTimeoutException: timeout
	at okhttp3.internal.http2.HttpOverHttp2Test.concurrentRequestWithEmptyFlowControlWindow(HttpOverHttp2Test.java:388)
[INFO]
[INFO] Results:
[INFO]
[ERROR] Errors:
[ERROR]   HttpOverHttp2Test.concurrentRequestWithEmptyFlowControlWindow:388 » SocketTimeout
[ERROR]   HttpOverHttp2Test.concurrentRequestWithEmptyFlowControlWindow:388 » SocketTimeout
[INFO]
[ERROR] Tests run: 90, Failures: 0, Errors: 2, Skipped: 5
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 52.072 s
[INFO] Finished at: 2018-07-09T10:03:16-04:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.21.0:test (default-test) on project okhttp-tests: There are test failures.
[ERROR]
[ERROR] Please refer to /Users/nroy/github/okhttp/okhttp-tests/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date]-jvmRun[N].dump, [date].dumpstream and [date]-jvmRun[N].dumpstream.

**Comment by natez0r** [comment]
I ran my test app with frame logging, here are the logs [frame_log.txt](https://github.com/square/okhttp/files/2177465/frame_log.txt)

Socket timeout was thrown at 5:59:24 in the logs

> 07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: Exception thrown:
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: com.google.android.exoplayer2.upstream.HttpDataSource$HttpDataSourceException: Unable to connect to https://video.twimg.com/ext_tw_video/974711634357874688/pr/pl/oJogd9M9-jDXIHQz.m3u8
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: 	at com.google.android.exoplayer2.ext.okhttp.OkHttpDataSource.open(OkHttpDataSource.java:171)
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: 	at test_app.nroy.twitter.com.exoplayerplayground.DownloadTest$3.load(DownloadTest.java:153)
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: 	at com.google.android.exoplayer2.upstream.Loader$LoadTask.run(Loader.java:317)
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: 	at java.lang.Thread.run(Thread.java:764)
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: Caused by: java.net.SocketTimeoutException: timeout
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: 	at okhttp3.internal.http2.Http2Stream$StreamTimeout.newTimeoutException(Http2Stream.java:605)
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: 	at okhttp3.internal.http2.Http2Stream$StreamTimeout.exitAndThrowIfTimedOut(Http2Stream.java:613)
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: 	at okhttp3.internal.http2.Http2Stream.takeResponseHeaders(Http2Stream.java:146)
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: 	at okhttp3.internal.http2.Http2Codec.readResponseHeaders(Http2Codec.java:127)
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: 	at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:88)
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: 	at test_app.nroy.twitter.com.exoplayerplayground.MainActivity.lambda$onCreate$0$MainActivity(MainActivity.java:29)
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: 	at test_app.nroy.twitter.com.exoplayerplayground.MainActivity$$Lambda$0.intercept(Unknown Source:22)
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: 	at com.facebook.stetho.okhttp3.StethoInterceptor.intercept(StethoInterceptor.java:56)
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: 	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:45)
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: 	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: 	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: 	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:126)
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: 	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:200)
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: 	at okhttp3.RealCall.execute(RealCall.java:77)
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: 	at com.google.android.exoplayer2.ext.okhttp.OkHttpDataSource.open(OkHttpDataSource.java:167)
07-09 17:59:24.591 23623 23857 E ExoPlayerPlayground: 	... 5 more

**Comment by dave-r12** [comment]
> When I wrote that comment, I had checked out your patch and that test failed for me:

I thought I had reproduced but realized I made a change to okhttp module for the test to pass (need to flush the HEADERS frame for the second request.) Can you compile/install that module (okhttp) and run again?

**Comment by dave-r12** [comment]
> I ran my test app with frame logging, here are the logs frame_log.txt

That's helpful, thanks. It almost looks like there is a buffer sitting between the client/server and it's trickling DATA frames to the client. In other words, there's a backlog of DATA frames that are waiting to be delivered from the first request and the response to the second request is at the end of that backlog. What happens if you increase the read timeout of the second request?

**Comment by natez0r** [comment]
> That's helpful, thanks. It almost looks like there is a buffer sitting between the client/server and it's trickling DATA frames to the client. In other words, there's a backlog of DATA frames that are waiting to be delivered from the first request and the response to the second request is at the end of that backlog. What happens if you increase the read timeout of the second request?

heh, I wonder if this is how genymotion throttles connections to make them 'Edge', which is not a real actual approximation of network speed.

**Comment by natez0r** [comment]
> I thought I had reproduced but realized I made a change to okhttp module for the test to pass (need to flush the HEADERS frame for the second request.) Can you compile/install that module (okhttp) and run again?

after doing a `mvn clean install` the tests pass for me

File: okhttp/src/main/java/okhttp3/internal/http2/Http2Stream.java [hunk 2]
```diff
@@ -180,9 +180,17 @@ public void sendResponseHeaders(List<Header> responseHeaders, boolean out) throw
...
+    // flow-control window is fully depleted.
+    boolean connectionWindowEmpty;
+    synchronized (connection) {
+      connectionWindowEmpty = connection.bytesLeftInWriteWindow == 0L;
+    }
+
     connection.writeSynReply(id, outFinished, responseHeaders);
 
-    if (outFinished) {
+    if (outFinished || connectionWindowEmpty) {
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/http2/Http2Stream.java]: Tempted to refactor this so it’s more symmetric with the code in `Http2Connection.newStream()`. In particular:
 - we could replace the `outFinished` local with a different boolean, `flushHeaders`.
 - we could only check if the connection window is empty if `flushHeaders` is false.

But as-is this definitely makes responses and requests work more similarly and therefore I think it’s a big improvement.

Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/http2/Http2Stream.java]: ie.
```
if (!flushHeaders) {
  synchronized (connection) {
    flushHeaders = connection.bytesLeftInWriteWindow == 0L;
  }
}
```

**Comment by swankjesse** [comment]
Nice improvement.

File: okhttp/src/main/java/okhttp3/internal/http2/Http2Stream.java (hunk 2 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/http2/Http2Stream.java]: Turns out it isn’t quite that simple, cause we need to keep outFinished to send with the headers. Regardless, I did this fixup.

**Pull request merged**